components/etcd.libsonnet

```diff
@@ -162,80 +161,102 @@ local k = import 'ksonnet/ksonnet.beta.4/k.libsonnet';
             --initial-cluster-state new \
             --data-dir /var/run/etcd/default.etcd
       |||,
-    }),
-
-  service:
-    local service = k.core.v1.service;
-    local ports = service.mixin.spec.portsType;
-
-    service.new(
-      etcd.config.name,
-      etcd.config.podLabelSelector,
-      [
-        ports.newNamed('client', 2379, 2379) +
-        ports.withProtocol('TCP'),
-        ports.newNamed('peer', 2380, 2380) +
-        ports.withProtocol('TCP'),
-      ]
-    ) +
-    service.mixin.metadata.withNamespace(etcd.config.namespace) +
-    service.mixin.metadata.withLabels(etcd.config.commonLabels),
+    },
+  },
+
+  service: {
+    apiVersion: 'v1',
+    kind: 'Service',
+    metadata: {
+      name: etcd.config.name,
+      namespace: etcd.config.namespace,
+      labels: etcd.config.commonLabels,
+    },
+    spec: {
+      ports: [
+        { name: 'client', targetPort: 2379, port: 2379, protocol: 'TCP' },
+        { name: 'peer', targetPort: 2380, port: 2380, protocol: 'TCP' },
+      ],
+      selector: etcd.config.podLabelSelector,
+    },
+  },
 
   statefulSet:
-    local statefulset = k.apps.v1.statefulSet;
-    local container = statefulset.mixin.spec.template.spec.containersType;
-    local envVar = container.envType;
-    local containerPort = container.portsType;
-    local containerVolumeMount = container.volumeMountsType;
-
-    local c =
-      container.new('etcd', etcd.config.image) +
-      container.withCommand([
+    local c = {
+      name: 'etcd',
+      image: etcd.config.image,
+      command: [
         '/bin/sh',
         '-ec',
         '/scripts/etcd-server.sh',
-      ]) + container.withEnv([
+      ],
+      env: [
         { name: 'ETCDCTL_API', value: '3' },
         { name: 'INITIAL_CLUSTER_SIZE', value: '%d' % etcd.config.replicas },
         { name: 'SET_NAME', value: etcd.config.name },
-      ]) + container.withPorts([
-        containerPort.newNamed(2379, 'client'),
-        containerPort.newNamed(2380, 'peer'),
-      ]) + container.mixin.resources.withRequests({
-        cpu: '100m',
-        memory: '128Mi',
-      }) + container.mixin.resources.withLimits({
-        cpu: '200m',
-        memory: '256Mi',
-      }) + container.withVolumeMounts([
-        containerVolumeMount.new('scripts', '/scripts'),
-        containerVolumeMount.new('storage', '/var/run/etcd'),
-      ]) + {
-        lifecycle: {
-          preStop: {
-            exec: {
-              command: [
-                '/bin/sh',
-                '-ec',
-                '/scripts/etcd-pre-stop.sh',
-              ],
-            },
+      ],
+      ports: [
+        { name: port.name, containerPort: port.port }
+        for port in etcd.service.spec.ports
+      ],
+      volumeMounts: [
+        {
+          name: 'scripts',
+          mountPath: '/scripts',
+        },
+        {
+          name: 'storage',
+          mountPath: '/var/run/etcd',
+        },
```

Comment:
These would be fine in just a single line each. Just as a comment for a next time. No need to update.